### PR TITLE
docs: comment stored procedures

### PR DESCRIPTION
**Pull request by jniles**
This PR documents some of the stored procedures with extensive
comments describing the scenarios and ways in which the stored
procedures ought to behave.  It clarifies how their usage for new
contributors as well.

Partially addresses #2147.

### CODE REVIEW

File: server/models/procedures/cash.sql [hunk 1]
```diff
@@ -1,12 +1,158 @@
...
+
+---------
+SCENARIOS
+---------
+
+For all the following scenarios, please consider a world with two currencies A
+and B.  The enterprise supports both currencies A and B. A's minimum monetary
+unit is 50, and B's minimum monetary unit is .25.  The enterprise's base
+currency is A.  Amounts expressed in "A" will be written 100.00A, and amounts
+express in "B" will be written 100.00B.
```
Comment (mbayopanda) [inline server/models/procedures/cash.sql:20]: 💃 this is very useful

File: server/models/procedures/cash.sql [hunk 2]
```diff
@@ -1,12 +1,158 @@
...
+that the debtor pay 10000A as a down payment.  The debtor chooses to pay in
+currency B at the cash window.  They will pay (10000A * (1B / 1000A)) = 10B.
+
+The debtor provides 10B as payment.  The application will record the following:
+ 1. A 10000A debit against the cashbox (representing 10B put into it)
+ 2. A 10000A credit for the debtor.
+
+In the metadata of the transaction, it is recorded that the debtor actually paid
+in B a value of 10B.  However, from the enterprise's perspective, they paid the
+equivalent in currency A.
```
Comment (mbayopanda) [inline server/models/procedures/cash.sql:57]: 👍

File: server/models/procedures/cash.sql [hunk 3]
```diff
@@ -1,12 +1,158 @@
...
+A debtor is invoiced for a product 750A, with an exchange rate of 1000A to 1B.
+If they choose to pay the invoice that day in currency B, they will be required
+to pay 0.75B.
+
+They go to the window to pay in currency B.  They pay 0.75B.  The following
+transaction is recorded:
+ 1. A 750A debit against the cashbox account
+ 2. A 750 credit against the debtor and their invoice.
+
+As above, the fact they paid with currency B is recorded as metadata.
```
Comment (mbayopanda) [inline server/models/procedures/cash.sql:72]: 👍

File: server/models/procedures/cash.sql [hunk 4]
```diff
@@ -1,12 +1,158 @@
...
+enterprise table's `gain_account_id`.  The resulting transaction will consist of
+three lines:
+ 1. A debit of 0.75B against the cashbox account.
+ 2. A credit of 0.68B to the debtor for their invoice.
+ 3. A credit of 0.7B against the gain account.
+
+NOTE: the values above are expressed in currency B for simplicity.  Since the
+enterprise is run in currency A, all those values will be converted into
+currency A with the exchange rate on the day it was paid.  In this case, 1100A
+to 1B.
```
Comment (mbayopanda) [inline server/models/procedures/cash.sql:105]: 👍

File: server/models/procedures/cash.sql [hunk 5]
```diff
@@ -1,12 +1,158 @@
...
+The debtor pays 1B, even though the total of their invoices was 1.09B.  The
+application will write a single transaction that consists of 4 lines:
+ 1. A debit of 1B against the cashbox account.
+ 2. A credit of 0.681B against the first invoice of the debtor
+ 3. A credit of 0.409B against the second invoice of the debtor
+ 4. A debit of 0.09B against the loss account.
+
+At the end of this procedure, the debtor will have equalized all their debts,
+and the loss account will have made up the difference to ensure that every
+invoice was correctly balanced.
```
Comment (mbayopanda) [inline server/models/procedures/cash.sql:142]: 👍 What degree of details, good job

File: server/models/procedures/invoicing.sql [hunk 6]
```diff
@@ -70,24 +96,45 @@ CREATE PROCEDURE StageSubsidy(
...
+    1) Billing Services place percentage increase on the invoice in proportion
+      to the base invoice cost.
+    2) Subsidies place a percentage reduction on the invoice in proportion to
+      the invoice cost.
+    3) Billing Services are applied first, then Subsidies are applied to the
+      adjusted invoice amount.
+
+  The final value of this algorithm is recorded in the invoices table as the
+  cost of the invoice.  In the posting journal, the invoice...
+*/
```
Comment (mbayopanda) [inline server/models/procedures/invoicing.sql:136]: 👍

File: server/models/procedures/invoicing.sql [hunk 7]
```diff
@@ -169,13 +250,20 @@ BEGIN
...
-  CALL PostingSetupUtil(date, enterprise_id, project_id, currency_id, current_fiscal_year_id, current_period_id, current_exchange_rate, enterprise_currency_id, transaction_id, gain_account_id, loss_account_id);
+  CALL PostingSetupUtil(
+    date, enterprise_id, project_id, currency_id, current_fiscal_year_id,
+    current_period_id, current_exchange_rate, enterprise_currency_id,
+    transaction_id, gain_account_id, loss_account_id
+  );
 
-  -- Check that all invoice items have sales accounts - if they do not the transaction will be imbalanced
+  -- Check that all invoice items have sales accounts - if they do not the
+  -- transaction will be unbalanced and the account_id will be NULL
```
Comment (mbayopanda) [inline server/models/procedures/invoicing.sql:266]: 👍

**Review by mbayopanda — APPROVED** [review]
That is really outstanding the way you have described procedures. Good job

**Comment by jniles** [comment]
:(.  Apparently there are some lethal comments in there.

**Pull request merged**